File: .credentials/get-credentials.py
# ABOUTME: Shared credential retrieval module.
# ABOUTME: Resolves credentials from env, a mounted secret file, or the host store.
#
# Usage:  from importlib.machinery import SourceFileLoader; mod = SourceFileLoader("gc", ".credentials/get-credentials.py").load_module()
# Inputs: Environment variables, /run/secrets/credentials.env, host credential store, credentials-map.json
# Outputs: Credential values as strings

from __future__ import annotations

import json
import os
import platform
import subprocess
from pathlib import Path

_CREDENTIALS_JSON = Path(__file__).resolve().parent / "credentials-map.json"
_MOUNTED_CREDENTIALS_FILE = Path(
    os.environ.get(
        "DEVCONTAINER_CREDENTIALS_FILE",
        "/run/secrets/credentials.env",
    )
)


def _load_credential_map() -> dict[str, str]:
    with open(_CREDENTIALS_JSON) as f:
        return json.load(f)


def _read_macos_secret(target: str) -> str:
    result = subprocess.run(
        ["security", "find-generic-password", "-s", target, "-w"],
        capture_output=True, text=True, check=False,
    )
    if result.returncode != 0:
        return ""
    return result.stdout.strip()


def _read_windows_secret(target: str) -> str:
    script = (
        "if (-not (Get-Command Get-StoredCredential -ErrorAction SilentlyContinue)) "
        "{ Write-Error 'Get-StoredCredential command not available. Install CredentialManager module.'; exit 1 }\n"
        f"$cred = Get-StoredCredential -Target '{target}'\n"
        f"if (-not $cred) {{ Write-Error \"Credential '{target}' not found in Windows Credential Manager.\"; exit 1 }}\n"
        "$cred.GetNetworkCredential().Password"
    )
    result = subprocess.run(
        ["powershell", "-NoProfile", "-Command", script],
        capture_output=True, text=True, check=False,
    )
    if result.returncode != 0:
        return ""
    return result.stdout.strip()


def _read_from_store(target: str) -> str:
    """Read a credential from the host OS credential store."""
    system = platform.system()
    if system == "Darwin":
        return _read_macos_secret(target)
    elif system == "Windows":
        return _read_windows_secret(target)
    return ""


def _read_mounted_credentials() -> dict[str, str]:
    """Read the Dev Container's mounted credential file, if present.

    The file uses Docker env-file syntax: one ``NAME=value`` entry per line.
    It is read afresh for every lookup so updates to the host-side bind mount
    are immediately visible inside the container.
    """
    try:
        lines = _MOUNTED_CREDENTIALS_FILE.read_text(encoding="utf-8").splitlines()
    except (OSError, UnicodeError):
        return {}

    credentials: dict[str, str] = {}
    for line in lines:
        stripped = line.strip()
        if not stripped or stripped.startswith("#") or "=" not in line:
            continue
        name, value = line.split("=", 1)
        name = name.strip()
        if name:
            credentials[name] = value
    return credentials


def get_credential(env_var: str, keychain_target: str) -> str:
    """Resolve a credential without placing it in the global container environment.

    Resolution order is environment, mounted secret file, then host credential
    store. Returns an empty string when the credential is unavailable.
    """
    value = os.environ.get(env_var, "")
    if value:
        return value
    value = _read_mounted_credentials().get(env_var, "")
    if value:
        return value
    return _read_from_store(keychain_target)


def get_all_credentials() -> tuple[dict[str, str], list[str]]:
    """Resolve all credentials defined in credentials-map.json.

    Returns (found, missing) where found is {env_var: value} and missing is
    a list of keychain targets that could not be resolved.
    """
    credential_map = _load_credential_map()
    found = {}
    missing = []
    for target, env_var in credential_map.items():
        value = get_credential(env_var, target)
        if value:
            found[env_var] = value
        else:
            missing.append(target)
    return found, missing

File: .credentials/get-credentials.R
# ABOUTME: Shared credential retrieval module for R.
# ABOUTME: Resolves credentials from env, a mounted secret file, or the host store.
#
# Run:     source(".credentials/get-credentials.R")
# Inputs:  Environment variables, /run/secrets/credentials.env, host credential
#          store, and credentials-map.json
# Outputs: Credential values returned by get_credential() or get_all_credentials()

.credentials_source_file <- tryCatch(
  normalizePath(sys.frame(1)$ofile, mustWork = TRUE),
  error = function(e) normalizePath(".credentials/get-credentials.R", mustWork = FALSE)
)
.credentials_dir <- dirname(.credentials_source_file)
.credentials_json <- file.path(.credentials_dir, "credentials-map.json")
.mounted_credentials_file <- Sys.getenv(
  "DEVCONTAINER_CREDENTIALS_FILE",
  unset = "/run/secrets/credentials.env"
)

.load_credential_map <- function() {
  jsonlite::fromJSON(.credentials_json, simplifyVector = TRUE)
}

.run_secret_command <- function(command, args) {
  output <- suppressWarnings(tryCatch(
    system2(command, args, stdout = TRUE, stderr = TRUE),
    error = function(e) structure(character(), status = 1L)
  ))
  status <- attr(output, "status", exact = TRUE)
  if (!is.null(status) && status != 0L) {
    return("")
  }
  trimws(paste(output, collapse = "\n"))
}

.read_macos_secret <- function(target) {
  .run_secret_command(
    "security",
    c("find-generic-password", "-s", shQuote(target), "-w")
  )
}

.read_windows_secret <- function(target) {
  escaped_target <- gsub("'", "''", target, fixed = TRUE)
  script <- paste0(
    "if (-not (Get-Command Get-StoredCredential -ErrorAction SilentlyContinue)) ",
    "{ Write-Error 'Get-StoredCredential command not available. Install CredentialManager module.'; exit 1 }\n",
    "$cred = Get-StoredCredential -Target '", escaped_target, "'\n",
    "if (-not $cred) { Write-Error \"Credential '", escaped_target,
    "' not found in Windows Credential Manager.\"; exit 1 }\n",
    "$cred.GetNetworkCredential().Password"
  )
  .run_secret_command(
    "powershell",
    c("-NoProfile", "-Command", shQuote(script))
  )
}

.read_from_store <- function(target) {
  system <- Sys.info()[["sysname"]]
  if (identical(system, "Darwin")) {
    return(.read_macos_secret(target))
  }
  if (identical(system, "Windows")) {
    return(.read_windows_secret(target))
  }
  ""
}

.read_mounted_credentials <- function() {
  lines <- tryCatch(
    readLines(.mounted_credentials_file, warn = FALSE, encoding = "UTF-8"),
    error = function(e) character()
  )
  credentials <- character()
  for (line in lines) {
    stripped <- trimws(line)
    if (!nzchar(stripped) || startsWith(stripped, "#") || !grepl("=", line, fixed = TRUE)) {
      next
    }
    separator <- regexpr("=", line, fixed = TRUE)[1]
    name <- trimws(substr(line, 1L, separator - 1L))
    if (nzchar(name)) {
      credentials[[name]] <- substr(line, separator + 1L, nchar(line))
    }
  }
  credentials
}

get_credential <- function(env_var, keychain_target) {
  value <- Sys.getenv(env_var, unset = "")
  if (nzchar(value)) {
    return(value)
  }
  mounted <- .read_mounted_credentials()
  value <- if (env_var %in% names(mounted)) mounted[[env_var]] else ""
  if (nzchar(value)) {
    return(value)
  }
  .read_from_store(keychain_target)
}

get_all_credentials <- function() {
  credential_map <- .load_credential_map()
  found <- character()
  missing <- character()
  for (target in names(credential_map)) {
    env_var <- unname(credential_map[[target]])
    value <- get_credential(env_var, target)
    if (nzchar(value)) {
      found[[env_var]] <- value
    } else {
      missing <- c(missing, target)
    }
  }
  list(found = found, missing = missing)
}
